flopy/mf6/utils/flopy_plugins/plugin_interface.py

```diff
@@ -0,0 +1,1008 @@
+        this method to execute any initialization code in your flopy plugin.
+
+        Parameters
+        ----------
+        fd_debug : file descriptor or None
+            Debug file descriptor
+
+        Returns
+        ----------
+        bool : Success/Failure
+        """
+        if fd_debug is not None:
+            self.fd_debug = fd_debug
+        return True
+
+    def sim_complete(self):
+        """This method is called by MFSimulation.run_simulation after the
+        simulation has completed.  Override this method to execute any
+        post-model processing or cleanup code in your flopy plugin.
+
+        Returns
+        ----------
+        bool : Success/Failure
+        """
+        return True
+
+    def val_from_cellid(self, cell_id, grid_array):
+        """This method retrieves a value from an array in the shape of the
+        model grid for a particular cell id.  Structured, vertex and
+        unstructured grids are supported.
+
+        Parameters
+        ----------
+        cell_id : tuple
+            Cell ID corresponding to the location in the array where a
+            value should be retrieved.
+        grid_array: numpy ndarray
+            Array containing value to be retrieved.
+
+        Returns
+        ----------
+        value : Value in a ndarray, None if grid type can not be resolved.
+        """
+        if self.mg.grid_type == "structured":
+            assert len(cell_id) == 3
+            return grid_array[cell_id[0], cell_id[1], cell_id[2]]
+        elif self.mg.grid_type == "vertex":
+            assert len(cell_id) == 2
+            return grid_array[cell_id[0], cell_id[1]]
+        elif self.mg.grid_type == "unstructured":
+            assert len(cell_id) == 1
+            return grid_array[cell_id[0]]
+        return None
+
+    @property
+    def dis_type(self):
+        """Returns a string specifying the type of dis package the model uses,
+        which can be used in the BMI interface.
+
+        Returns
+        ----------
+        str : The type of dis package used by the model
+        """
+        if self._dis_type is None:
+            dis_pkg = self.model.get_package("dis")
+            self._dis_type = dis_pkg._package_type
+        return self._dis_type
+
+    def get_node(self, cell_id):
+        """Given a cell ID return the corresponding model node.
+
+        Parameters
+        ----------
+        cell_id : tuple
+            Cell ID
+
+        Returns
+        ----------
+        int : Node corresponding to cell_id in the model.
+        """
+        if self.mg.grid_type == "structured":
+            assert len(cell_id) == 3
+            return self.mg.get_node([cell_id])[0]
+        elif self.mg.grid_type == "vertex":
+            assert len(cell_id) == 2
+            return self.mg.get_node([cell_id])[0]
+        else:
+            assert len(cell_id) == 1
+            return cell_id[0] + 1
+
+    @staticmethod
+    def sq_saturation(top, bot, x, c1=None, c2=None):
```

Comment:
Maybe there is good reason for this, but this does not seem like a good place for replicated low-level GWF NPF calculation routines.